FWCore/Framework/interface/stream/ThinningProducer.h

```diff
@@ -21,11 +22,59 @@
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 
 #include <memory>
+#include <optional>
+#include <type_traits>
 
 namespace edm {
 
   class EventSetup;
 
+  namespace detail {
+    template <typename T>
+    struct IsStdOptional {
+      static constexpr bool value = false;
+    };
+    template <typename T>
+    struct IsStdOptional<std::optional<T>> {
+      static constexpr bool value = true;
+    };
+
+    // by default a linear container
+    template <typename Collection, typename = std::void_t<>>
+    struct ElementType {
```

Comment:
Should this be handled by a traits class instead? That way you should be able to get explicit specialization for DetSetVector. As it stands now this would also fire for `std::vector<std::unique_ptr<std::vector<int>>`.

Reply:
I agree, a traits class looks cleaner.